source/MaterialXGraphEditor/Graph.cpp

```diff
@@ -2505,187 +2515,206 @@ void Graph::addLink(ed::PinId inputPinId, ed::PinId outputPinId)
         return;
     }
 
-    if (inputPin->_connected == false)
+    // Perform kind check
+    bool kindsMatch = (outputPin->_kind == inputPin->_kind);
+    if (kindsMatch)
     {
-        int upNode = getNodeId(inputPinId);
-        int downNode = getNodeId(outputPinId);
-        UiNodePtr uiDownNode = _graphNodes[downNode];
-        UiNodePtr uiUpNode = _graphNodes[upNode];
-        if (!uiDownNode || !uiUpNode)
-        {
-            ed::RejectNewItem();
-            return;
-        }
+        ed::RejectNewItem();
+        showLabel("Invalid connection due to same input/output kind", ImColor(50, 50, 50, 255));
+        return;
+    }
 
-        // make sure there is an implementation for node
-        const mx::ShaderGenerator& shadergen = _renderer->getGenContext().getShaderGenerator();
+    int upNode = getNodeId(outputPinId);
+    int downNode = getNodeId(inputPinId);
+    UiNodePtr uiDownNode = _graphNodes[downNode];
+    UiNodePtr uiUpNode = _graphNodes[upNode];
+    if (!uiDownNode || !uiUpNode)
+    {
+        ed::RejectNewItem();
+        return;
+    }
 
-        // Prevent direct connecting from input to output
-        if (uiDownNode->getInput() && uiUpNode->getOutput())
-        {
-            ed::RejectNewItem();
-            showLabel("Direct connections between inputs and outputs is invalid", ImColor(50, 50, 50, 255));
-            return;
-        }
+    // make sure there is an implementation for node
+    const mx::ShaderGenerator& shadergen = _renderer->getGenContext().getShaderGenerator();
 
-        // Find the implementation for this nodedef if not an input or output uinode
-        if (uiDownNode->getInput() && _isNodeGraph)
+    // Prevent direct connecting from input to output
+    if (uiDownNode->getInput() && uiUpNode->getOutput())
+    {
+        ed::RejectNewItem();
+        showLabel("Direct connections between inputs and outputs is invalid", ImColor(50, 50, 50, 255));
+        return;
+    }
+
+    // Find the implementation for this nodedef if not an input or output uinode
+    if (uiDownNode->getInput() && _isNodeGraph)
+    {
+        ed::RejectNewItem();
+        showLabel("Cannot connect to inputs inside of graph", ImColor(50, 50, 50, 255));
+        return;
+    }
+    else if (uiUpNode->getNode())
+    {
+        mx::ShaderNodeImplPtr impl = shadergen.getImplementation(*_graphNodes[upNode]->getNode()->getNodeDef(), _renderer->getGenContext());
+        if (!impl)
         {
             ed::RejectNewItem();
-            showLabel("Cannot connect to inputs inside of graph", ImColor(50, 50, 50, 255));
+            showLabel("Invalid Connection: Node does not have an implementation", ImColor(50, 50, 50, 255));
             return;
         }
-        else if (uiUpNode->getNode())
+    }
+
+    if (ed::AcceptNewItem())
+    {
+        // If the accepting node already has a link, remove it
+        if (inputPin->_connected)
```

Comment:
This is the crux of the new work - if we do have a connection for our input pin, we remove the existing link associated with it.